### ISSUE TITLE
When migration fails, restore the DB Schema to the "stable" state before try migrating up.

### ISSUE THREAD

**Issue by Raffaello** [doctrine/DoctrineMigrationsBundle#113]
When migrating up, if an exception is thrown, some changes in the DB schema are made and not restored to the previous state. It means that some migration commands ran succesfully and alter the schema and other not. 

It require that if an exception it is thrown, run a "back propagation" to restore the succesful sql command to the previous state with the migration down. Otherwise We have a DB Schema at middle way from one migration to another and It is in an inconsistency state. 

It is required to manually restore to the previous state before run the unsucessful migration.

To do it automatically, it should be required that the migration up and down have the inverse/same order of the line, i.e the 1st line of up have is counterpart in the last or first (depend on the algorithm policy) of line of down.

It is required to store the line in an array, when an exception throw run the counterpart starting from the correspective counterpart of the last succesful line. (they could be easily stored in a stack, but anyway it is not required because they are sorted)

[if it is the same order just downcount until zero index, otherwise start from the index until the end. (with a stack just run until the stack is empy.]

How to reprodoce the error (just an example):
Migration**** failed during Execution. Error An exception occurred while executing 'ALTER TABLE table ADD CONSTRAINT FK_53DE51E73A302744 FOREIGN KEY (table_table2_id) REFERENCES table2 (table2_id)':

SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`db_name`.`#sql-2717_af20`, CONSTRAINT `FK_53DE51E73A302744` FOREIGN KEY (`table2_table2_id`) REFERENCES `table2` (`table2`))

**Comment by mikeSimonson** [comment]
This is a doctrine migration issue, but I can already tell you that it can not be fixed.

All queries in a migration are run in a transaction so that if there is a problem, it revert automatically to a sane state.

But not all the database behave the same way when you make DDL queries (queries that change the database structure by adding a column or an index for example).

Some will auto commit the transaction (it's probably what happened to you), other will be able to do those change in a transaction and so on.

The only solution you have to avoid that can of issue is to separate your DDL statement in an other migration file.

Good luck

**Comment by stof** [comment]
the other solution is to use a database supporting DDL statements in transactions (Postgres FTW :smile: )

**Comment by Raffaello** [comment]
Thanks,
I see...
It is already a right good solution with transaction. 
Unfortunately it is not at all supported in my current schema.
For sure I've got some non transactional table for performance reason. Good to know.
  
But anyway, If I have to split manually the file, in this case is EXACT what I proposed for the migration up/down rollback operation. Just run single statements at time. 
I think it is not at all a "high priority feature" even if it is still implementable. Make no sense at all due to the transaction.

Yes, i will do "smaller migration" . :)

Thanks a lot.
